core/src/storage2/lazy/lazy_imap.rs

```diff
@@ -0,0 +1,733 @@
+    /// - Use [`LazyIndexMap::put`]`(None)` in order to remove an element.
+    /// - Prefer this method over [`LazyIndexMap::put_get`] if you are not interested
+    ///   in the old value of the same cell index.
+    ///
+    /// # Panics
+    ///
+    /// - If the lazy chunk is in an invalid state that forbids interaction.
+    /// - If the decoding of the old element at the given index failed.
+    pub fn put(&mut self, index: Index, new_value: Option<V>) {
+        self.entries_mut()
+            .insert(index, Box::new(Entry::new(new_value, EntryState::Mutated)));
+    }
+}
+
+impl<V> SpreadLayout for LazyIndexMap<V>
+where
+    V: PackedLayout,
+{
+    const FOOTPRINT: u64 = 1_u64 << 32;
+
+    fn pull_spread(ptr: &mut KeyPtr) -> Self {
+        Self::lazy(ptr.next_for::<Self>())
+    }
+
+    fn push_spread(&self, ptr: &mut KeyPtr) {
+        let offset_key = ptr.next_for::<Self>();
+        for (&index, entry) in self.entries().iter() {
+            let root_key = offset_key + index;
+            entry.push_packed_root(&root_key);
+        }
+    }
+
+    #[inline]
+    fn clear_spread(&self, _ptr: &mut KeyPtr) {
+        // Low-level lazy abstractions won't perform automated clean-up since
+        // they generally are not aware of their entire set of associated
+        // elements. The high-level abstractions that build upon them are
+        // responsible for cleaning up.
+    }
+}
+
+impl<V> LazyIndexMap<V>
+where
+    V: PackedLayout,
+{
+    /// Clears the underlying storage of the entry at the given index.
+    ///
+    /// # Safety
+    ///
+    /// For performance reasons this does not synchronize the lazy index map's
+    /// memory-side cache which invalidates future accesses the cleared entry.
+    /// Care should be taken when using this API.
+    ///
+    /// The general use of this API is to streamline `Drop` implementations of
+    /// high-level abstractions that build upon this low-level data strcuture.
+    pub fn clear_packed_at(&self, index: Index) {
+        let root_key = self.key_at(index).expect("cannot clear in lazy state");
+        if <V as SpreadLayout>::REQUIRES_DEEP_CLEAN_UP {
+            // We need to load the entity before we remove its associated contract storage
+            // because it requires a deep clean-up which propagates clearing to its fields,
+            // for example in the case of `T` being a `storage::Box`.
+            let entity = self.get(index).expect("cannot clear a non existing entity");
+            clear_packed_root::<V>(&entity, &root_key);
+        } else {
+            // The type does not require deep clean-up so we can simply clean-up
+            // its associated storage cell and be done without having to load it first.
+            crate::env::clear_contract_storage(root_key);
+        }
+    }
+}
+
+impl<V> LazyIndexMap<V>
+where
+    V: PackedLayout,
+{
+    /// Returns an offset key for the given index.
+    pub fn key_at(&self, index: Index) -> Option<Key> {
```

Comment:
Is it worth refactoring out the duplication with `LazyHashHap`? The main difference being this method

Reply:
There are actually many more differences but in theory you are right. I actually even started with a shared code base and just separated their implementations because it became a bit too complex. However, we can think about some follow-up PRs that take on this code dupe in isolation.